File: DSA/DSA04025 DÃY SỐ VÔ HẠN.cpp
#include <bits/stdc++.h>
using namespace std;

const long long MOD = 1e9 + 7;

void multiply(vector<vector<long long>> &A, vector<vector<long long>> &B)
{
    vector<vector<long long>> C(2, vector<long long>(2));

    C[0][0] = ((A[0][0]*B[0][0])%MOD + (A[0][1]*B[1][0])%MOD)%MOD;
    C[0][1] = ((A[0][0]*B[0][1])%MOD + (A[0][1]*B[1][1])%MOD)%MOD;

    C[1][0] = ((A[1][0]*B[0][0])%MOD + (A[1][1]*B[1][0])%MOD)%MOD;
    C[1][1] = ((A[1][0]*B[0][1])%MOD + (A[1][1]*B[1][1])%MOD)%MOD;

    A[0][0] = C[0][0];
    A[0][1] = C[0][1];
    A[1][0] = C[1][0];
    A[1][1] = C[1][1];
}

vector<vector<long long>> power(vector<vector<long long>> A, int times)
{
    vector<vector<long long>> ans ={{1, 0}, {0, 1}};
    while (times)
    {
        if (times & 1)
        {
            multiply(ans, A);
        }
        multiply(A, A);
        times >>= 1;
    }
    
    return ans;
}

void submit()
{
    int n;
    cin >> n;

    vector<vector<long long>> F =   {{1, 0},
                                    {0, 0}};
    vector<vector<long long>> M =   {{1, 1}, 
                                    {1, 0}};

    vector<vector<long long>> ans = power(M, n-1);

    multiply(ans, F);

    cout << ans[0][0]%MOD;

    cout << endl;
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int test;
    cin >> test;
    while (test--)
    {
        submit();
    }
    

    return 0;
}

File: DSA_python/DSA04025 DÃY SỐ VÔ HẠN.py
def submit():
    n = int(input())
    F = [[1, 0],
         [0, 0]]
    M = [[1, 1],
         [1, 0]]
    
    def multiply(A, B):
        C = [[0]*2 for _ in range(2)]
        for i in range(2):
            for j in range(2):
                for k in range(2):
                    C[i][j] = (C[i][j] + A[i][k]*B[k][j])%MOD
        return C

    def power(A, times):
        ans = [[0]*2 for _ in range(2)]
        for i in range(2):
            for j in range(2):
                if i == j:
                    ans[i][j] = 1
        
        while times:
            if times&1:
                ans = multiply(ans, A)
            A = multiply(A, A)
            times >>= 1
        return ans

    ans = power(M, n-1)
    ans = multiply(ans, F)

    print(ans[0][0]%MOD)


MOD = 10**9+7

test = int(input())
for _ in range(test):
    submit()
